gradle.properties

```diff
@@ -21,7 +21,7 @@ org.gradle.caching=true
 org.gradle.vfs.watch=true
 
 # Uncomment following line to temporary bypass dependencies verification
-#org.gradle.dependency.verification=lenient
+org.gradle.dependency.verification=lenient
```

Comment:
https://github.com/gradle/gradle/issues/20234

Reply:
_(Whops this got a bit long, hope it helps readers who take the time to read to learn something new 😅)_

You can undo this if you add this to `build-logic/build.gradle.kts`:
```
buildscript {
    repositories {
        mavenCentral()
        //gradlePluginPortal()
    }
}
```

Or to be more specific:
```
buildscript {
    repositories {
        // Note: this is actually redundant, because exclusiveContent wins anyway, but it makes things
        // a bit more explicit and future-proof, in case the exclusiveContent gets removed.
        gradlePluginPortal {
            content {
                // See https://github.com/gradle/gradle/issues/20234
                excludeVersion("de.undercouch", "gradle-download-task", "4.1.1")
            }
        }
        // Note: with exclusiveContent, order doesn't matter.
        exclusiveContent {
            forRepository { mavenCentral() }
            filter {
                // See https://github.com/gradle/gradle/issues/20234
                includeVersion("de.undercouch", "gradle-download-task", "4.1.1")
            }
        }
    }
}
```

---

Here's a bit of an explanation:
As shown in https://github.com/gradle/gradle/issues/20234#issuecomment-1254096198, the poms are different in the different repositories. It seems that mavenCentral() PGP-validates for this artifact's POM, so that's the main source.

The validation error you're seeing here:
```
> Configure project :build-logic
Dependency verification failed for configuration ':build-logic:classpath'
One artifact failed verification: gradle-download-task-4.1.1.pom (de.undercouch:gradle-download-task:4.1.1) from repository Gradle Central Plugin Repository
This can indicate that a dependency has been compromised. Please carefully verify the signatures and checksums.
```
can be interpreted as follows:
 * `Configure project :build-logic` -> so the problem is before executing settings.gradle.kts from the main project
 * `configuration  ':build-logic:classpath'` -> so the problem is with the `buildscript`'s classpath inside the `build-logic` module. Remember the old way we added plugins (`buildscript { dependencies { classpath ...`)
 * `gradle-download-task-4.1.1.pom from repository Gradle Central Plugin Repository` -> the problem is with this artifact in `gradlePluginPortal()`
 * via `gradlew -p build-logic :buildEnvironment` I can see:
    ```
    > Task :buildEnvironment

    ------------------------------------------------------------
    Root project 'build-logic'
    ------------------------------------------------------------

    classpath
    +--- org.gradle.kotlin.kotlin-dsl:org.gradle.kotlin.kotlin-dsl.gradle.plugin:2.3.3
    |    \--- org.gradle.kotlin:gradle-kotlin-dsl-plugins:2.3.3
    |         +--- org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.21
    |         |    +--- de.undercouch:gradle-download-task:4.1.1
    ```
    So this is the path to the dependency transitively.
 * `org.gradle.kotlin.kotlin-dsl.gradle.plugin` is on `classpath` because
   * ``plugins { `kotlin-dsl` }``
   * if there's no `buildscript {}` in a build.gradle.kts
     * it either comes from `pluginManagement` in settings.gradle (which doesn't exist in build-logic)
     * or it is implied to be `buildscript { repositories { gradlePluginPortal() } }`
   * and of course `plugins {}` adds dependencies to `classpath` configuration

So to prevent a verification error we need to make sure the dependency comes from the right repository (mavenCentral()). We have multiple ways to achieve this:
 1) By explicitly declaring `mavenCentral() + gradlePluginPortal()` we're giving priority to `mavenCentral()` (which has the [right POM](https://github.com/gradle/gradle/issues/20234#issuecomment-1254096198), and for anything not present in central, we fall back to Gradle's. 
 2) By explicitly declaring `mavenCentral()` without `gradlePluginPortal()` we're restoring the original artifact resolution and therefore getting the right POM.
 3) By keeping `gradlePluginPortal()` as the main repository, but taking this specific artifact from `mavenCentral()` via `exclusiveContent`.

Relying on the order (1) feels a bit hacky, that's why (2) is probably preferable, but if the complexity is acceptable (pending pinterest's approval) maybe (3) is also ok because it's closer to the default Gradle behavior only fixing the one small thing that's broken.

---

As an additional note, we should be also mindful that `build-logic` gets built based on this `classpath`, but then its `runtimeClasspath` becomes part the including build's `pluginManagement` (`buildscript`>`classpath`). Which means that `pluginManagement.repositories` must apply the same fix (it already has both repos in the right order). At this point I tried using only `gradlePluginPortal()` in `/settings.gradle.kts` and it resulted in these validation errors:
```
> Configure project :
Dependency verification failed for configuration ':classpath'
3 artifacts failed verification:
  - gradle-download-task-4.1.1.pom (de.undercouch:gradle-download-task:4.1.1) from repository Gradle Central Plugin Repository
  - kotlin-gradle-plugin-1.7.10.module (org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.10) from repository Gradle Central Plugin Repository
  - org.jetbrains.kotlin.jvm.gradle.plugin-1.7.10.pom (org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin:1.7.10) from repository Gradle Central Plugin Repository

> Configure project :ktlint
Dependency verification failed for configuration ':ktlint:classpath'
4 artifacts failed verification:
  - gradle-download-task-4.1.1.pom (de.undercouch:gradle-download-task:4.1.1) from repository Gradle Central Plugin Repository
  - kotlin-gradle-plugin-1.7.10.module (org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.10) from repository Gradle Central Plugin Repository
  - org.jetbrains.kotlin.jvm.gradle.plugin-1.7.10.pom (org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin:1.7.10) from repository Gradle Central Plugin Repository
  - dokka-gradle-plugin-1.7.10.pom (org.jetbrains.dokka:dokka-gradle-plugin:1.7.10) from repository Gradle Central Plugin Repository
```

Reply:
But seems I fix the sum by 1dc4a305ff30bad7f5c133ae7fb42841f802ecd4.

Reply:
This is because a change in [buildSrc causes the whole project to become out-of-date](https://docs.gradle.org/7.2/userguide/organizing_gradle_projects.html#sec:build_sources).

Reply:
> https://github.com/pinterest/ktlint/commit/1dc4a305ff30bad7f5c133ae7fb42841f802ecd4 means you accept that the POMs are different (because you validated there's no funky stuff going on in there).

Would prefer to follow what Gradle did, and `buildscript { mavenCentral() }` might be a little noisy, all repos should be configured in settings.gradle might be neat.

Reply:
I meant that you changed the repository declarations too, not just the file structure. I agree with settings.gradle, but there's two of them and only one declares repos. The point here is replacing the repos with a trusted source.

I'm aware of the buildSrc problem, I did multiple migrations like this, and I'm doing a similar migration at work, but I couldn't prove where this solution is better. Do you know a surefire way to test this benefit?

Reply:
https://github.com/gradle/gradle/issues/21271#issuecomment-1191143416